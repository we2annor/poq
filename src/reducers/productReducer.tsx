import _ from "lodash";
import {
  FetchProductAction,
  DeleteProductAction,
  FETCH_PRODUCTS,
  DELETE_PRODUCT,
} from "../Actions/types";

export interface ProductState {
  products: Product[];
}

export interface Product {
  productId: number;
  name: string;
  description: string;
  price: number;
  priceWas: number;
  available: boolean;
  quantity: number;
  lowOnStock: boolean;
  promotionBadge: string;
  imageUrl: string;
}
const initialState: ProductState = {
  products: [],
};

export const productReducer = (
  state: ProductState = initialState,
  action: FetchProductAction | DeleteProductAction
) => {
  switch (action.type) {
    case FETCH_PRODUCTS:
      return { ...state.products, ..._.mapKeys(action.payload, "productId") };
    case DELETE_PRODUCT:
      const newProducts = Object.values(state);
      const filteredList = newProducts.filter(
        (product) => product.productId !== action.payload
      );
      return {
        ...filteredList,
      };
    default:
      return state;
  }
};
