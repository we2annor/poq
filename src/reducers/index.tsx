import { combineReducers } from "redux";
import { productReducer } from "./productReducer";

export type RootState = ReturnType<typeof rootReducer>;
export const rootReducer = combineReducers({ products: productReducer });
