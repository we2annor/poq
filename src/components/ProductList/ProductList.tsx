import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { fetchProduct, deleteProduct } from "../../Actions/index";
import { ProductState, Product } from "../../reducers/productReducer";
import ProductItem from "../Product/Product";

interface Props {
  products: Product[];
  fetchProduct: () => void;
  deleteProduct: (id: number) => void;
}

const ProductList: React.FC<Props> = ({
  products,
  fetchProduct,
  deleteProduct,
}) => {
  const [selected, setSelected] = useState(0);
  const [selectedProductList, setSelectedProductList] = useState<string[]>([]);

  useEffect(() => {
    const loadProducts = () => {
      fetchProduct();
    };

    loadProducts();
  }, [fetchProduct]);

  const getSelectedProducts = (product: HTMLInputElement) => {
    const { id, checked } = product;
    if (checked) {
      if (selectedProductList.indexOf(id) === -1) {
        setSelected(selected + 1);
        setSelectedProductList((selectedProductList) => [
          ...selectedProductList,
          id,
        ]);
      }
    } else {
      if (selectedProductList.indexOf(id) !== -1) {
        setSelected(selected - 1);
        setSelectedProductList(
          selectedProductList.filter((product) => product !== id)
        );
      }
    }
  };

  const onRemoveButtonClicked = () => {
    setSelectedProductList([]);
    setSelected(0);
    return selectedProductList.map((product) =>
      deleteProduct(parseInt(product))
    );
  };

  return (
    <>
      {selected > 0 && (
        <button
          className='product-list__selected-btn'
          onClick={onRemoveButtonClicked}
        >{`Remove ${selected > 0 ? selected : ""} selected product${
          selected > 1 ? "s" : ""
        }`}</button>
      )}

      <div className='product-list'>
        {products.map((product, index) => (
          <div key={index}>
            <ProductItem
              getSelectedProducts={getSelectedProducts}
              product={product}
            />
          </div>
        ))}
      </div>
    </>
  );
};

const matchPropsToState = (state: ProductState) => {
  return {
    products: Object.values(state.products),
  };
};

export default connect(matchPropsToState, { fetchProduct, deleteProduct })(
  ProductList
);
