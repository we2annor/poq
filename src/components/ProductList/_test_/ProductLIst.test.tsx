import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import renderer from "react-test-renderer";
import { rootReducer } from "../../../reducers";
import ProductList from "../ProductList";

afterEach(cleanup);

// const observe = jest.fn();

// window.IntersectionObserver = jest.fn(function () {
//   this.observe = observe;
// });

describe("When ProductList Mounts", () => {
  test("Shoult render successfully", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    render(
      <Provider store={store}>
        <ProductList />
      </Provider>
    );
  });

  test("Matches snapshot", () => {
    const store = createStore(rootReducer, applyMiddleware(thunk));
    const tree = renderer.create(
      <Provider store={store}>
        <ProductList />
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });
});
