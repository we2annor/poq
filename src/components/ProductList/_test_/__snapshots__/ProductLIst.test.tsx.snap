// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When Address Mounts Matches snapshot 1`] = `
<div
  className="product-list"
>
  <div>
    <div
      className="product"
    >
      <span
        className="product__checkBox"
      >
        <input
          checked={false}
          className="remove-checkbox"
          id=""
          name=""
          onChange={[Function]}
          type="checkbox"
        />
      </span>
      <div
        className="product__img-wrapper"
      >
        <img
          className="sc-bdfBwQ iBymIc"
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
        />
      </div>
      <div
        className="product__info"
      >
        <div
          className="product__info--name"
        />
        <div
          className="product__info--price"
        >
          £
           
          <span>
            £
          </span>
        </div>
        <div>
          
        </div>
        <span>
          <span
            className="out-of-stock"
          >
            Out of stock
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`When ProductList Mounts Matches snapshot 1`] = `
<div
  className="product-list"
>
  <div>
    <div
      className="product"
    >
      <span
        className="product__checkBox"
      >
        <input
          checked={false}
          className="remove-checkbox"
          id=""
          name=""
          onChange={[Function]}
          type="checkbox"
        />
      </span>
      <div
        className="product__img-wrapper"
      >
        <img
          className="sc-bdfBwQ iBymIc"
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkqAcAAIUAgUW0RjgAAAAASUVORK5CYII="
        />
      </div>
      <div
        className="product__info"
      >
        <div
          className="product__info--name"
        />
        <div
          className="product__info--price"
        >
          £
           
          <span>
            £
          </span>
        </div>
        <div>
          
        </div>
        <span>
          <span
            className="out-of-stock"
          >
            Out of stock
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
